File: java/advanced/InlineAdaptiveBannerExample/app/src/main/java/com/google/android/gms/example/inlineadaptivebannerexample/RecyclerViewAdapter.java
package com.google.android.gms.example.inlineadaptivebannerexample;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.google.android.gms.ads.AdView;
import java.util.List;

/**
 * The {@link RecyclerViewAdapter} class.
 * <p>The adapter provides access to the items in the {@link MenuItemViewHolder}
 * or the {@link AdViewHolder}.</p>
 */
class RecyclerViewAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    // A menu item view type.
    private static final int MENU_ITEM_VIEW_TYPE = 0;

    // The banner ad view type.
    private static final int BANNER_AD_VIEW_TYPE = 1;

    // An Activity's Context.
    private final Context context;

    // The list of banner ads and menu items.
    private final List<Object> recyclerViewItems;

    /**
     * For this example app, the recyclerViewItems list contains only
     * {@link MenuItem} and {@link AdView} types.
     */
    public RecyclerViewAdapter(Context context, List<Object> recyclerViewItems) {
        this.context = context;
        this.recyclerViewItems = recyclerViewItems;
    }

  /**
   * The {@link MenuItemViewHolder} class. Provides a reference to each view in the menu item view.
   */
  public static class MenuItemViewHolder extends RecyclerView.ViewHolder {
    private final TextView menuItemName;
    private final TextView menuItemDescription;
    private final TextView menuItemPrice;
    private final TextView menuItemCategory;
    private final ImageView menuItemImage;

        MenuItemViewHolder(View view) {
            super(view);
            menuItemImage = view.findViewById(R.id.menu_item_image);
            menuItemName = view.findViewById(R.id.menu_item_name);
            menuItemPrice = view.findViewById(R.id.menu_item_price);
            menuItemCategory = view.findViewById(R.id.menu_item_category);
            menuItemDescription = view.findViewById(R.id.menu_item_description);
        }
    }

  /** The {@link AdViewHolder} class. */
  public static class AdViewHolder extends RecyclerView.ViewHolder {

        AdViewHolder(View view) {
            super(view);
        }
    }

    @Override
    public int getItemCount() {
        return recyclerViewItems.size();
    }

    /**
     * Determines the view type for the given position.
     */
    @Override
    public int getItemViewType(int position) {
        return (position % MainActivity.ITEMS_PER_AD == 0) ? BANNER_AD_VIEW_TYPE
                : MENU_ITEM_VIEW_TYPE;
    }

  /**
   * Creates a new view for a menu item view or a banner ad view based on the viewType. This method
   * is invoked by the layout manager.
   */
  @NonNull
  @Override
  public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int viewType) {
        switch (viewType) {
            case MENU_ITEM_VIEW_TYPE:
                View menuItemLayoutView = LayoutInflater.from(viewGroup.getContext()).inflate(
                        R.layout.menu_item_container, viewGroup, false);
                return new MenuItemViewHolder(menuItemLayoutView);
            case BANNER_AD_VIEW_TYPE:
                // fall through
            default:
                View bannerLayoutView = LayoutInflater.from(
                        viewGroup.getContext()).inflate(R.layout.banner_ad_container,
                        viewGroup, false);
                return new AdViewHolder(bannerLayoutView);
        }
    }

  /**
   * Replaces the content in the views that make up the menu item view and the banner ad view. This
   * method is invoked by the layout manager.
   */
  @Override
  public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        int viewType = getItemViewType(position);
        switch (viewType) {
            case MENU_ITEM_VIEW_TYPE:
                MenuItemViewHolder menuItemHolder = (MenuItemViewHolder) holder;
                MenuItem menuItem = (MenuItem) recyclerViewItems.get(position);

                // Get the menu item image resource ID.
                String imageName = menuItem.getImageName();
                int imageResID = context.getResources().getIdentifier(imageName, "drawable",
                        context.getPackageName());

                // Add the menu item details to the menu item view.
                menuItemHolder.menuItemImage.setImageResource(imageResID);
                menuItemHolder.menuItemName.setText(menuItem.getName());
                menuItemHolder.menuItemPrice.setText(menuItem.getPrice());
                menuItemHolder.menuItemCategory.setText(menuItem.getCategory());
                menuItemHolder.menuItemDescription.setText(menuItem.getDescription());
                break;
            case BANNER_AD_VIEW_TYPE:
                // fall through
            default:
                AdViewHolder bannerHolder = (AdViewHolder) holder;
                AdView adView = (AdView) recyclerViewItems.get(position);
                ViewGroup adCardView = (ViewGroup) bannerHolder.itemView;
                // The AdViewHolder recycled by the RecyclerView may be a different
                // instance than the one used previously for this position. Clear the
                // AdViewHolder of any subviews in case it has a different
                // AdView associated with it, and make sure the AdView for this position doesn't
                // already have a parent of a different recycled AdViewHolder.
                if (adCardView.getChildCount() > 0) {
                    adCardView.removeAllViews();
                }
                if (adView.getParent() != null) {
                    ((ViewGroup) adView.getParent()).removeView(adView);
                }

                // Add the banner ad to the ad view.
                adCardView.addView(adView);
        }
    }

}

File: kotlin/advanced/InlineAdaptiveBannerExample/app/src/main/java/com/google/android/gms/example/inlineadaptivebannerexample/RecyclerViewAdapter.kt
package com.google.android.gms.example.inlineadaptivebannerexample

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import com.google.android.gms.ads.AdView
import com.google.android.gms.example.inlineadaptivebannerexample.RecyclerViewAdapter.AdViewHolder
import com.google.android.gms.example.inlineadaptivebannerexample.RecyclerViewAdapter.MenuItemViewHolder

/**
 * The [RecyclerViewAdapter] class.
 *
 * The adapter provides access to the items in the [MenuItemViewHolder] or the [AdViewHolder].
 */
internal class RecyclerViewAdapter
/** For this example app, the recyclerViewItems list contains only [MenuItem] and [AdView] types. */
(
  // An Activity's Context.
  private val context: Context,
  // The list of banner ads and menu items.
  private val recyclerViewItems: List<Any>,
) : RecyclerView.Adapter<ViewHolder>() {
  /** The [MenuItemViewHolder] class. Provides a reference to each view in the menu item view. */
  inner class MenuItemViewHolder internal constructor(view: View) : ViewHolder(view) {
    val menuItemName: TextView
    val menuItemDescription: TextView
    val menuItemPrice: TextView
    val menuItemCategory: TextView
    val menuItemImage: ImageView

    init {
      menuItemImage = view.findViewById(R.id.menu_item_image)
      menuItemName = view.findViewById(R.id.menu_item_name)
      menuItemPrice = view.findViewById(R.id.menu_item_price)
      menuItemCategory = view.findViewById(R.id.menu_item_category)
      menuItemDescription = view.findViewById(R.id.menu_item_description)
    }
  }

  /** The [AdViewHolder] class. */
  inner class AdViewHolder internal constructor(view: View?) : ViewHolder(view!!)

  override fun getItemCount(): Int {
    return recyclerViewItems.size
  }

  /** Determines the view type for the given position. */
  override fun getItemViewType(position: Int): Int {
    return if (position % MainActivity.ITEMS_PER_AD == 0) BANNER_AD_VIEW_TYPE
    else MENU_ITEM_VIEW_TYPE
  }

  /**
   * Creates a new view for a menu item view or a banner ad view based on the viewType. This method
   * is invoked by the layout manager.
   */
  override fun onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): ViewHolder {
    return when (viewType) {
      MENU_ITEM_VIEW_TYPE -> {
        val menuItemLayoutView =
          LayoutInflater.from(viewGroup.context)
            .inflate(R.layout.menu_item_container, viewGroup, false)
        MenuItemViewHolder(menuItemLayoutView)
      }
      else -> {
        val bannerLayoutView =
          LayoutInflater.from(viewGroup.context)
            .inflate(R.layout.banner_ad_container, viewGroup, false)
        AdViewHolder(bannerLayoutView)
      }
    }
  }

  /**
   * Replaces the content in the views that make up the menu item view and the banner ad view. This
   * method is invoked by the layout manager.
   */
  override fun onBindViewHolder(holder: ViewHolder, position: Int) {
    when (getItemViewType(position)) {
      MENU_ITEM_VIEW_TYPE -> {
        val menuItemHolder = holder as MenuItemViewHolder
        val menuItem = recyclerViewItems[position] as MenuItem

        // Get the menu item image resource ID.
        val imageName = menuItem.imageName
        val imageResID = context.resources.getIdentifier(imageName, "drawable", context.packageName)

        // Add the menu item details to the menu item view.
        menuItemHolder.menuItemImage.setImageResource(imageResID)
        menuItemHolder.menuItemName.text = menuItem.name
        menuItemHolder.menuItemPrice.text = menuItem.price
        menuItemHolder.menuItemCategory.text = menuItem.category
        menuItemHolder.menuItemDescription.text = menuItem.description
      }
      else -> {
        val bannerHolder = holder as AdViewHolder
        val adView = recyclerViewItems[position] as AdView
        val adCardView = bannerHolder.itemView as ViewGroup
        // The AdViewHolder recycled by the RecyclerView may be a different
        // instance than the one used previously for this position. Clear the
        // AdViewHolder of any subviews in case it has a different
        // AdView associated with it, and make sure the AdView for this position doesn't
        // already have a parent of a different recycled AdViewHolder.
        if (adCardView.childCount > 0) {
          adCardView.removeAllViews()
        }
        if (adView.parent != null) {
          (adView.parent as ViewGroup).removeView(adView)
        }

        // Add the banner ad to the ad view.
        adCardView.addView(adView)
      }
    }
  }

  companion object {
    // A menu item view type.
    private const val MENU_ITEM_VIEW_TYPE = 0

    // The banner ad view type.
    private const val BANNER_AD_VIEW_TYPE = 1
  }
}
